src/Storages/MeiliSearch/SourceMeiliSearch.cpp

```diff
@@ -0,0 +1,119 @@
+#include "SourceMeiliSearch.h"
+#include <base/JSON.h>
+#include <base/range.h>
+#include "Common/Exception.h"
+#include "Common/quoteString.h"
+#include "Columns/ColumnString.h"
+#include "Columns/ColumnVector.h"
+#include "Columns/IColumn.h"
+#include "Core/Field.h"
+#include "base/types.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int MEILISEARCH_EXCEPTION;
+}
+
+MeiliSearchSource::MeiliSearchSource(
+    const MeiliSearchConfiguration & config,
+    const Block & sample_block,
+    UInt64 max_block_size_,
+    std::unordered_map<String, String> query_params_)
+    : SourceWithProgress(sample_block.cloneEmpty())
+    , connection(config)
+    , max_block_size{max_block_size_}
+    , query_params{query_params_}
+    , offset{0}
+{
+    description.init(sample_block);
+
+    String columns_to_get = "[";
+    for (const auto & col : description.sample_block)
+        columns_to_get += doubleQuoteString(col.name) + ",";
+    
+    columns_to_get.back() = ']';
+
+    query_params[doubleQuoteString("attributesToRetrieve")] = columns_to_get;
+    query_params[doubleQuoteString("limit")] = std::to_string(max_block_size);
+}
+
+
+MeiliSearchSource::~MeiliSearchSource() = default;
+
+void insertWithTypeId(MutableColumnPtr & column, JSON kv_pair, ExternalResultDescription::ValueType type_id)
+{
+    if (type_id == ExternalResultDescription::ValueType::vtUInt64 || 
+        type_id == ExternalResultDescription::ValueType::vtUInt32 || 
+        type_id == ExternalResultDescription::ValueType::vtUInt16 || 
+        type_id == ExternalResultDescription::ValueType::vtUInt8)
+    {
+        auto value = kv_pair.getValue().get<UInt64>();
+        column->insert(value);
+    }
+    else if (type_id == ExternalResultDescription::ValueType::vtInt64 || 
+             type_id == ExternalResultDescription::ValueType::vtInt32 || 
+             type_id == ExternalResultDescription::ValueType::vtInt16 || 
+             type_id == ExternalResultDescription::ValueType::vtInt8)
+    {
+        auto value = kv_pair.getValue().get<Int64>();
+        column->insert(value);
+    }
+    else if (type_id == ExternalResultDescription::ValueType::vtString)
+    {
+        auto value = kv_pair.getValue().get<String>();
+        column->insert(value);
+    }
+    else if (type_id == ExternalResultDescription::ValueType::vtFloat64 || 
+             type_id == ExternalResultDescription::ValueType::vtFloat32)
+    {
+        auto value = kv_pair.getValue().get<Float64>();
+        column->insert(value);
+    }
```

Comment:
In the last commit, there will be all supported types, except for null, boolean, and json objects, since it is not clear how to store them in clickhouse.

Reply:
null -- just wrap data type into nullable, see there is a nullabity field https://github.com/ClickHouse/ClickHouse/blob/master/src/Core/ExternalResultDescription.h#L42 in external result description.
for bool -- UInt8 or maybe take a look at https://github.com/ClickHouse/ClickHouse/pull/31072.
json -- can we just convert it to srting for now?